### PR TITLE
[AIRFLOW-6162] Add back serialization as a module

### PR DESCRIPTION
**Pull request by potiuk**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6162

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/serialization/serialized_dag.py [hunk 1]
```diff
@@ -101,7 +102,7 @@ def deserialize_dag(cls, encoded_dag: dict) -> "SerializedDAG":
         setattr(dag, 'full_filepath', dag.fileloc)
         for task in dag.task_dict.values():
             task.dag = dag
-            serializable_task: SerializedBaseOperator = task
+            serializable_task: SerializedBaseOperator = cast(SerializedBaseOperator, task)
```
Comment (kaxil) [inline airflow/serialization/serialized_dag.py]: Is this required?

Comment (potiuk) [reply airflow/serialization/serialized_dag.py]: Yes. MyPy complains if it's not there. Maybe it's worth looking whether there is a good reason for that @kaxil ?

Comment (potiuk) [reply airflow/serialization/serialized_dag.py]: I will shortly

**Review by ashb — APPROVED** [review]

File: airflow/serialization/serialized_dag.py (hunk 1 above)
Comment (kaxil) [reply airflow/serialization/serialized_dag.py]: Cool, other than that I am fine with this PR :)

Thanks @potiuk

**Review by kaxil — APPROVED** [review]

**Review by nemkin — APPROVED** [review]
Tried this locally and it fixed the missing serialization module after running pip install . Thank you Jarek :)

**Review by mik-laj — APPROVED** [review]

File: airflow/serialization/serialized_dag.py (hunk 1 above)
Comment (potiuk) [reply airflow/serialization/serialized_dag.py]: I think it's worth looking at it @kaxil -> there are still some parts of the code that needs some more typing to catch potential errors (DagBag especially) but this one looks a bit fishy to me.

Comment (potiuk) [reply airflow/serialization/serialized_dag.py]: I think the root cause is that .subdag field is only set in SerializedBaseOperator and that's why the type is forced here.

**Comment by kaxil** [comment]
Yes unfortunately, subdag field is only set in 2 places. 1) In
SubDagOperator and 2) In SerializedBaseOperator.

**Comment by potiuk** [comment]
@kaxil -> I have a fix in making for that (but the #6596 should be merge first because we have the errors fixed in this PR).

**Comment by potiuk** [comment]
In my upcoming fix I also merge serialized_dag and serialized_baseoperator into one module 'serialized_objects' - because there are cyclic dependencies (again detected by pylint) between those two classes as well and it makes perfect sense to group them in one module.

**Comment by potiuk** [comment]
@kaxil @ashb @mik-laj -> Just pushed fixup which removes the cast and fixes it "properly" I think. I moved subdag to BaseOperator and added few missing type annotations. This has let to more cyclic dependencies (between SerializedDag and SerializedBaseOperator). I believe those two classes are so closely coupled that it makes sense to put them in one module (basically that's what pylint is complaining about when telling about "cyclic imports" - that classes in separate modules are so closely coupled that they are referring each other).

BTW. I think the more type annotations we add the more couplings we detect and there will be another set of splits/merges between the core classes. We have still quite a few of not-annotated classes that are hiding such couplings.

**Review by turbaszek — APPROVED** [review]

File: airflow/serialization/base_serialization.py [hunk 2]
```diff
@@ -154,8 +154,8 @@ def _serialize(cls, var):  # pylint: disable=too-many-return-statements
         (3) Operator has a special field CLASS to record the original class
             name for displaying in UI.
         """
-        from airflow.serialization.serialized_dag import SerializedDAG
-        from airflow.serialization.serialized_baseoperator import SerializedBaseOperator
+        from airflow.serialization.serialized_objects import SerializedDAG
+        from airflow.serialization.serialized_objects import SerializedBaseOperator
```
Comment (kaxil) [inline airflow/serialization/base_serialization.py]:
```suggestion
        from airflow.serialization.serialized_objects import SerializedDAG, SerializedBaseOperator
```

File: airflow/serialization/base_serialization.py [hunk 3]
```diff
@@ -207,8 +207,8 @@ def _serialize(cls, var):  # pylint: disable=too-many-return-statements
     @classmethod
     def _deserialize(cls, encoded_var):  # pylint: disable=too-many-return-statements
         """Helper function of depth first search for deserialization."""
-        from airflow.serialization.serialized_dag import SerializedDAG
-        from airflow.serialization.serialized_baseoperator import SerializedBaseOperator
+        from airflow.serialization.serialized_objects import SerializedDAG
+        from airflow.serialization.serialized_objects import SerializedBaseOperator
```
Comment (kaxil) [inline airflow/serialization/base_serialization.py]:
```suggestion
        from airflow.serialization.serialized_objects import SerializedDAG, SerializedBaseOperator
```

**Comment by potiuk** [comment]
I will rebase it on top of #6596  to see if it works with hose fixes.

File: airflow/serialization/base_serialization.py (hunk 3 above)
Comment (potiuk) [reply airflow/serialization/base_serialization.py]: Yep. Isort will ... sort it out ...

File: airflow/serialization/base_serialization.py (hunk 2 above)
Comment (potiuk) [reply airflow/serialization/base_serialization.py]: Yeah. I sort will take care of it :)

**Comment by potiuk** [comment]
Rebased now on top of #6596

**Comment by potiuk** [comment]
And some more pylint complaints (rightfully so) I also moved BaseSerialization in the same "serialization_objects.py". Now this module is really what it should be :).

**Comment by potiuk** [comment]
@ashb @kaxil -> pls take a look. It's changed since the approval a lot.

**Pull request merged**